OpenRA.Mods.Common/Traits/Air/Aircraft.cs

```diff
@@ -799,10 +799,22 @@ public void SetPosition(Actor self, CPos cell, SubCell subCell = SubCell.Any)
 
 		public void SetPosition(Actor self, WPos pos)
 		{
-			CenterPosition = pos;
-
 			if (!self.IsInWorld)
+			{
+				CenterPosition = pos;
 				return;
+			}
+
+			var wasAirbourne = self.World.Map.DistanceAboveTerrain(CenterPosition).Length >= Info.MinAirborneAltitude;
```

Comment:
We have a functions called for these events. `OnAirborneAltitudeReached()` and `OnAirborneAltitudeLeft()`.
We should reuse hem

Reply:
It's difficult because `RemoveInfluence` needs to be called with the old data first. If we have already updated the position then we've missed our chance to call `RemoveInfluence`. If you have an elegant suggestion that accounts for that I'm open to it.

Reply:
We don't miss it, see line 824

Reply:
Oh, I understand. A comment should be added explaining that influence should be added / removed before changing CenterPosition?

Also DistanceAboveTerrain calculation should be reused in line 821, or more accurately the rest of the code should take advantage of our new these new checks, as now there's a lot of redundant code in SetPosition function

Reply:
It also feels incorrect. If the aircraft was nudged to the side (by some code) while ascending influence won't be updated correctly either way. And if it was not nudged, the aircraft will still be in the same cell, thus not mattering if `RemoveInfluence` was called before or after setting CenterPosition

Reply:
Also, according to #20312 shouldn't we set CenterPosition before changing influence?

Reply:
> Oh, I understand. A comment should be added explaining that influence should be added / removed before changing CenterPosition?

All the other implementations do this without a comment, so I'm fairly meh on it.

> Also DistanceAboveTerrain calculation should be reused in line 821, or more accurately the rest of the code should take advantage of our new these new checks, as now there's a lot of redundant code in SetPosition function

I feel like this might be somewhat dangerous, since we're mutating the variable in between and it could be confusing and/or wrong. I feel it's safer to repeat the calculation for clarity.

> It also feels incorrect. If the aircraft was nudged to the side (by some code) while ascending influence won't be updated correctly either way. And if it was not nudged, the aircraft will still be in the same cell, thus not mattering if `RemoveInfluence` was called before or after setting CenterPosition

Do you have a concrete example? I feel like any such external code would need to call `Aircraft.SetPosition` which would therefore ensure influence is dealt with.

> Also, according to #20312 shouldn't we set CenterPosition before changing influence?

For Aircraft, setting the `CenterPosition` updates the `TopLeft` at the same time, as `TopLeft` is defined in terms of the `CenterPosition`. Therefore when `AddInfluence` is called it does see updated values for both, which is what we want.

Reply:
> Do you have a concrete example? I feel like any such external code would need to call Aircraft.SetPosition which would therefore ensure influence is dealt with.
> For Aircraft, setting the CenterPosition updates the TopLeft at the same time, as TopLeft is defined in terms of the CenterPosition. Therefore when AddInfluence is called it does see updated values for both, which is what we want.

I don't understand why we want to remove influence before changing `CenterPosition`. The only thing that changes is height and does `RemoveInfluence` care about height? If not then there's no point is calling it before a change. `RemoveInfluence` could be called right before `AddInfluence`.

If height doesn't matter then I can only assume that this change is meant to `RemoveInfluence` from the spot the aircraft was landed on by calling `RemoveInfluence` while ascending. 

In which case the position of the aircraft shouldn't have changed regardless and if it has via moving the landed aircraft from cell A to cell B via outside script like chronoshift then the check will still be woefully incorrect, as it doesn't remove the influence from the position where chrono was called but where it currently is. 

Which means that the script should have been handled at the moment of chrono and not when ascending. And we end up at square one where calling `RemoveInfluence` does the same thing wether it's called before or after  changing vertical position 

> I feel like this might be somewhat dangerous, since we're mutating the variable in between and it could be confusing and/or wrong. I feel it's safer to repeat the calculation for clarity.

In our case reusing would be smart as literally no additional behaviours are triggered when changing `CenterPosition`. But sure, you can future proof this by keeping redundant code incase we add additional logic to the `CenterPosition` setter

Reply:
We call `RemoveInfluence` which removes influence based on the current list of `OccupiedCells`. When we set `CenterPosition`, the `OccupiedCells` could change. This occurs because `Aircraft.OccupiedCells` has two branches. We return different lists based on the condition `self.World.Map.DistanceAboveTerrain(CenterPosition).Length >= Info.MinAirborneAltitude`.

Then we call `AddInfluence` which will add the new influence according to `OccupiedCells`.

So we are forced to:
- Remove influence before updating the CenterPosition. If we don't do this we might not remove the influence we had previously applied.
- Add influence after updating the CenterPosition. If we don't do this we might apply the wrong influence.

Reply:
> Remove influence before updating the CenterPosition. If we don't do this we might not remove the influence we had previously applied.

Then the check is still wrong because the aircraft can have a different `CenterPosition` assigned to it before ascending above or descending below MinAirborneAltitude. We need to apply this every time `CenterPosition` is changed while the unit has influence, instead on the MinAirborneAltitude threshold

Reply:
And landingCells updated to the new position as well

> All the other implementations do this without a comment, so I'm fairly meh on it.

Then the other implementations should have a comment added as well. It seems fairly hidden that the `OccupiedCells` returns different values depending on `MinAirborneAltitude`. A comment like this should be added
```C#
// OccupiedCells returns a different value depending on aircraft being above or below MinAirborneAltitude. 
// We catch this transition and update ActorMap accordingly